File: packages/sdk-php/src/PhoneAPI.php
<?php

declare(strict_types=1);

namespace Melaya;

/**
 * Phone API — pair and control Android devices connected to the Melaya runner.
 *
 * Maps to /api/v1/private/phone/*. Agents use these endpoints to drive a
 * paired phone (tap, type, read screen state, launch apps) via the Melaya APK.
 *
 * @example
 * ```php
 * $sdk = new Melaya\Melaya(apiKey: getenv('MK'));
 *
 * $pair = $sdk->phone->pair();
 * echo $pair['code']; // show this on screen; user enters it in the APK
 *
 * $devices = $sdk->phone->listDevices();
 * $tree = $sdk->phone->screenTree();
 * ```
 */
class PhoneAPI
{
    public function __construct(private readonly HttpClient $http) {}

    /**
     * Start phone device pairing.
     * Returns a pairing code to enter on the Melaya APK.
     */
    public function pair(): array
    {
        return $this->http->post('/api/v1/private/phone/pair');
    }

    /** List all paired phone devices for the authenticated user. */
    public function listDevices(): array
    {
        $response = $this->http->get('/api/v1/private/phone/devices');
        return $response['devices'] ?? [];
    }

    /** Revoke a paired phone device by ID. */
    public function revokeDevice(string $deviceId): array
    {
        return $this->http->delete('/api/v1/private/phone/devices/' . rawurlencode($deviceId));
    }

    /** Get the current accessibility tree from the paired phone's screen. */
    public function screenTree(): array
    {
        return $this->http->get('/api/v1/private/phone/screen-tree');
    }

    /** List installed apps on the paired phone. */
    public function listApps(): array
    {
        $response = $this->http->get('/api/v1/private/phone/apps');
        return $response['result']['apps'] ?? [];
    }

    /**
     * Set the allowlist of apps that agents are permitted to interact with.
     *
     * @param string[] $packageNames List of Android package names (e.g. ['com.twitter.android'])
     */
    public function setAllowedApps(array $packageNames): array
    {
        $apps = array_map(static fn (string $package): array => ['package' => $package], $packageNames);
        return $this->http->put('/api/v1/private/phone/apps/allowed', ['apps' => $apps]);
    }

    /** Register the currently active pipeline run on the phone (used by agents). */
    public function registerActiveRun(string $runId): array
    {
        return $this->http->post('/api/v1/private/phone/active-run', ['runId' => $runId]);
    }
}

File: packages/sdk-java/src/main/java/org/melaya/PhoneAPI.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.List;
import java.util.Map;

/**
 * Phone API — pair and control Android devices connected to the Melaya runner.
 *
 * <p>Maps to {@code /api/v1/private/phone/*}. Agents use these endpoints to
 * drive a paired phone (tap, type, read screen state, launch apps) via the
 * Melaya APK.
 *
 * @example
 * <pre>{@code
 * JsonNode pair   = melaya.phone().pair();
 * String code     = pair.get("code").asText();
 * JsonNode devices = melaya.phone().listDevices();
 * JsonNode tree    = melaya.phone().screenTree();
 * }</pre>
 */
public class PhoneAPI {

    private final HttpClient http;

    PhoneAPI(HttpClient http) {
        this.http = http;
    }

    /**
     * Start phone device pairing — generates a pairing code to enter on the
     * Melaya APK.
     */
    public JsonNode pair() {
        return http.post("/api/v1/private/phone/pair", null);
    }

    /** List all paired phone devices for the authenticated user. */
    public JsonNode listDevices() {
        return http.get("/api/v1/private/phone/devices", null).path("devices");
    }

    /**
     * Revoke a paired phone device by ID.
     *
     * @param deviceId the device ID to revoke
     */
    public JsonNode revokeDevice(String deviceId) {
        return http.delete("/api/v1/private/phone/devices/" + deviceId, null);
    }

    /** Get the current accessibility tree from the paired phone's screen. */
    public JsonNode screenTree() {
        return http.get("/api/v1/private/phone/screen-tree", null);
    }

    /** List installed apps on the paired phone. */
    public JsonNode listApps() {
        return http.get("/api/v1/private/phone/apps", null).path("result").path("apps");
    }

    /**
     * Set the allowlist of apps that agents are permitted to interact with.
     *
     * @param packageNames Android package names to permit
     */
    public JsonNode setAllowedApps(List<String> packageNames) {
        List<Map<String, String>> apps = packageNames.stream()
                .map(packageName -> Map.of("package", packageName))
                .toList();
        return http.put("/api/v1/private/phone/apps/allowed", Map.of("apps", apps));
    }

    /**
     * Register the currently active pipeline run on the phone.
     * Used by agent pipelines that control the phone.
     *
     * @param runId the pipeline run ID
     */
    public JsonNode registerActiveRun(String runId) {
        return http.post("/api/v1/private/phone/active-run",
                MarketAPI.params("runId", runId));
    }
}
